File: scheduler/core/src/main/java/io/orkes/conductor/scheduler/service/SchedulerBulkServiceImpl.java
package io.orkes.conductor.scheduler.service;

import java.util.List;

import org.springframework.context.annotation.Conditional;
import org.springframework.stereotype.Service;

import com.netflix.conductor.annotations.Audit;
import com.netflix.conductor.annotations.Trace;
import com.netflix.conductor.common.model.BulkResponse;

import io.orkes.conductor.scheduler.config.SchedulerConditions;
import lombok.extern.slf4j.Slf4j;

@Audit
@Trace
@Service
@Slf4j
@Conditional(SchedulerConditions.class)
public class SchedulerBulkServiceImpl implements SchedulerBulkService {

    private final SchedulerService schedulerService;

    public SchedulerBulkServiceImpl(SchedulerService schedulerService) {
        this.schedulerService = schedulerService;
    }

    /**
     * Pause the list of schedules.
     *
     * @param scheduleNames - list of schedule names to perform pause operation on
     * @return bulk response object containing a list of succeeded schedules and a list of failed
     *     ones with errors
     */
    @Override
    public BulkResponse pauseSchedules(List<String> scheduleNames) {
        BulkResponse bulkResponse = new BulkResponse();

        for (String scheduleName : scheduleNames) {
            try {
                schedulerService.pauseSchedule(scheduleName);
                bulkResponse.appendSuccessResponse(scheduleName);
                log.debug("Successfully paused schedule: {}", scheduleName);
            } catch (Exception e) {
                log.error(
                        "bulk pauseSchedule exception, scheduleName {}, message: {}",
                        scheduleName,
                        e.getMessage(),
                        e);
                bulkResponse.appendFailedResponse(scheduleName, e.getMessage());
            }
        }

        log.info(
                "Bulk pause schedules completed. Success: {}, Failed: {}",
                bulkResponse.getBulkSuccessfulResults().size(),
                bulkResponse.getBulkErrorResults().size());

        return bulkResponse;
    }

    /**
     * Resume the list of schedules.
     *
     * @param scheduleNames - list of schedule names to perform resume operation on
     * @return bulk response object containing a list of succeeded schedules and a list of failed
     *     ones with errors
     */
    @Override
    public BulkResponse resumeSchedules(List<String> scheduleNames) {
        BulkResponse bulkResponse = new BulkResponse();

        for (String scheduleName : scheduleNames) {
            try {
                schedulerService.resumeSchedule(scheduleName);
                bulkResponse.appendSuccessResponse(scheduleName);
                log.debug("Successfully resumed schedule: {}", scheduleName);
            } catch (Exception e) {
                log.error(
                        "bulk resumeSchedule exception, scheduleName {}, message: {}",
                        scheduleName,
                        e.getMessage(),
                        e);
                bulkResponse.appendFailedResponse(scheduleName, e.getMessage());
            }
        }

        log.info(
                "Bulk resume schedules completed. Success: {}, Failed: {}",
                bulkResponse.getBulkSuccessfulResults().size(),
                bulkResponse.getBulkErrorResults().size());

        return bulkResponse;
    }
}

File: scheduler/core/src/main/typescript/io/orkes/conductor/scheduler/service/SchedulerBulkServiceImpl.ts
import { SchedulerBulkService } from './SchedulerBulkService';
import { SchedulerService } from './SchedulerService';
import { BulkResponse } from '../../../../../../../../../mock';

export class SchedulerBulkServiceImpl implements SchedulerBulkService {
    private readonly schedulerService: SchedulerService;
    private static readonly log = console;

    constructor(schedulerService: SchedulerService) {
        this.schedulerService = schedulerService;
    }

    /**
     * Pause the list of schedules.
     *
     * @param scheduleNames - list of schedule names to perform pause operation on
     * @return bulk response object containing a list of succeeded schedules and a list of failed
     *     ones with errors
     */
    public pauseSchedules(scheduleNames: string[]): BulkResponse {
        const bulkResponse = new BulkResponse();

        for (const scheduleName of scheduleNames) {
            try {
                this.schedulerService.pauseSchedule(scheduleName);
                bulkResponse.appendSuccessResponse(scheduleName);
                SchedulerBulkServiceImpl.log.debug(`Successfully paused schedule: ${scheduleName}`);
            } catch (e: any) {
                SchedulerBulkServiceImpl.log.error(
                    `bulk pauseSchedule exception, scheduleName ${scheduleName}, message: ${e.message}`,
                    e
                );
                bulkResponse.appendFailedResponse(scheduleName, e.message);
            }
        }

        SchedulerBulkServiceImpl.log.info(
            `Bulk pause schedules completed. Success: ${bulkResponse.getBulkSuccessfulResults().length}, Failed: ${bulkResponse.getBulkErrorResults().length}`
        );

        return bulkResponse;
    }

    /**
     * Resume the list of schedules.
     *
     * @param scheduleNames - list of schedule names to perform resume operation on
     * @return bulk response object containing a list of succeeded schedules and a list of failed
     *     ones with errors
     */
    public resumeSchedules(scheduleNames: string[]): BulkResponse {
        const bulkResponse = new BulkResponse();

        for (const scheduleName of scheduleNames) {
            try {
                this.schedulerService.resumeSchedule(scheduleName);
                bulkResponse.appendSuccessResponse(scheduleName);
                SchedulerBulkServiceImpl.log.debug(`Successfully resumed schedule: ${scheduleName}`);
            } catch (e: any) {
                SchedulerBulkServiceImpl.log.error(
                    `bulk resumeSchedule exception, scheduleName ${scheduleName}, message: ${e.message}`,
                    e
                );
                bulkResponse.appendFailedResponse(scheduleName, e.message);
            }
        }

        SchedulerBulkServiceImpl.log.info(
            `Bulk resume schedules completed. Success: ${bulkResponse.getBulkSuccessfulResults().length}, Failed: ${bulkResponse.getBulkErrorResults().length}`
        );

        return bulkResponse;
    }
}
